# Assignment 1


## 1 $n$-gram Language Model
**Q1**: Expand the above definition of $ p(\vec{w})$ using naive estimates of the parameters, such as $  p(w_4 \mid w_2, w_3) \stackrel{\tiny{\mbox{def}}}{=}  \frac{C(w_2~w_3~w_4)}{C(w_2~w_3)} $ where \( C(w_2 w_3 w_4) \) denotes the count of times the trigram $ w_2 w_3 w_4 $ was observed in a training corpus.

**A1**: Expand the above definition of $ p(\vec{w})$  using naive estimates of the parameters

\begin{align}
p(\vec{w}) \stackrel{\tiny{\mbox{def}}}{=} & p(w_1) \cdot p(w_2 | w_1) \cdot p(w_3 | w_1 w_2) \cdot p(w_4 | w_2 w_3) \cdots p(w_n | w_{n-2} w_{n-1}) \\
&= \frac{c(w_1)}{n} \cdot \frac{c(w_1 w_2)}{c(w_1)} \cdot \frac{c(w_1 w_2 w_3)}{c(w_1 w_2)} \cdot \frac{c(w_2 w_3 w_4)}{c(w_2 w_3)} \cdots \frac{c(w_{n-2} w_{n-1} w_n)}{c(w_{n-2} w_{n-1})}
\end{align}



**Q2**: One could also define a kind of reversed trigram language model $p_{reversed}$ that instead assumed the words were generated in reverse order (from right to left):
\begin{align} p_{reversed}(\vec{w}) \stackrel{\tiny{\mbox{def}}}{=}&p(w_n) \cdot p(w_{n-1} \mid w_n) \cdot p(w_{n-2} \mid w_{n-1} w_n) \cdot p(w_{n-3} \mid w_{n-2} w_{n-1}) \\ &\cdots p(w_2 \mid w_3 w_4) \cdot p(w_1 \mid w_2 w_3) \end{align}
By manipulating the notation, show that the two models are identical, i.e., $ p(\vec{w}) = p_{reversed}(\vec{w}) $ for any $ \vec{w} $ provided that both models use MLE parameters estimated from the same training data (see Q1 above).

**A2**:
show that the two models are identical:


Original Trigram Language Model
\begin{align}
p(\vec{w}) \stackrel{\tiny{\mbox{def}}}{=} & p(w_1) \cdot p(w_2 | w_1) \cdot p(w_3 | w_1 w_2) \cdot p(w_4 | w_2 w_3) \cdots p(w_n | w_{n-2} w_{n-1}) \\
&= \frac{c(w_1)}{n} \cdot \frac{c(w_1 w_2)}{c(w_1)} \cdot \frac{c(w_1 w_2 w_3)}{c(w_1 w_2)} \cdot \frac{c(w_2 w_3 w_4)}{c(w_2 w_3)} \cdots \frac{c(w_{n-2} w_{n-1} w_n)}{c(w_{n-2} w_{n-1})}
\end{align}

After fraction reduction,

\begin{align}
p(\vec{w}) &= \frac{c(w_1 w_2 w_3)}{n} \cdot  \frac{c(w_2 w_3 w_4)}{c(w_2 w_3)} \cdots \frac{c(w_{n-2} w_{n-1} w_n)}{c(w_{n-2} w_{n-1})}
\end{align}

Reversed Trigram Language Model:

Express the reversed trigram language model using maximum likelihood estimation as follows.

\begin{align}
p_{reversed}(\vec{w}) \stackrel{\tiny{\mbox{def}}}{=} & p(w_n) \cdot p(w_{n-1} | w_n) \cdot p(w_{n-2} | w_{n-1} w_n) \cdot p(w_{n-3} | w_{n-2} w_{n-1}) \cdots p(w_2 | w_3 w_4) \cdot p(w_1 | w_2 w_3) \\
&= \frac{c(w_n)}{n} \cdot \frac{c(w_{n-1} w_n)}{c(w_n)} \cdot \frac{c(w_{n-2} w_{n-1} w_n)}{c(w_{n-1} w_n)} \cdot \frac{c(w_{n-3} w_{n-2} w_{n-1})}{c(w_{n-2} w_{n-1})} \cdots \frac{c(w_2 w_3 w_4)}{c(w_3 w_4)} \cdot \frac{c(w_1 w_2 w_3)}{c(w_2 w_3)}
\end{align}

After fraction reduction,

\begin{align}
p_{reversed}(\vec{w}) &= \frac{c(w_{n-2} w_{n-1} w_n)}{n} \cdot \frac{c(w_{n-3} w_{n-2} w_{n-1})}{c(w_{n-2} w_{n-1})} \cdots \frac{c(w_2 w_3 w_4)}{c(w_3 w_4)} \cdot \frac{c(w_1 w_2 w_3)}{c(w_2 w_3)} \\
 &= \frac{c(w_1 w_2 w_3)}{n} \cdot  \frac{c(w_2 w_3 w_4)}{c(w_2 w_3)} \cdots \frac{c(w_{n-2} w_{n-1} w_n)}{c(w_{n-2} w_{n-1})}\\
 &= p(\vec{w})
\end{align}

Through the above maximum likelihood estimation, it is proved that $ p(\vec{w}) = p_{reversed}(\vec{w}) $ .


## 2 $N$-gram Language Model Implementation

In [ ]:
!wget -O train.txt https://raw.githubusercontent.com/qtli/COMP7607-Fall2023/master/assignments/A1/data/lm/train.txt
!wget -O dev.txt https://raw.githubusercontent.com/qtli/COMP7607-Fall2023/master/assignments/A1/data/lm/dev.txt
!wget -O test.txt https://raw.githubusercontent.com/qtli/COMP7607-Fall2023/master/assignments/A1/data/lm/test.txt

--2023-10-22 13:52:11--  https://raw.githubusercontent.com/qtli/COMP7607-Fall2023/master/assignments/A1/data/lm/train.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 6640478 (6.3M) [text/plain]
Saving to: ‘train.txt’

train.txt           100%[===================>]   6.33M  --.-KB/s    in 0.1s    

2023-10-22 13:52:11 (62.4 MB/s) - ‘train.txt’ saved [6640478/6640478]

--2023-10-22 13:52:11--  https://raw.githubusercontent.com/qtli/COMP7607-Fall2023/master/assignments/A1/data/lm/dev.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Le

### 2.1 Building vocabulary

**Code**

In [ ]:
import collections
import re

# read train data
with open('train.txt', 'r', encoding='utf-8') as train_file:
    train_data = train_file.readlines()

# read dev data
with open('dev.txt', 'r', encoding='utf-8') as dev_file:
    dev_data = dev_file.readlines()

words_counts = collections.Counter()

for sentence in train_data:
    words = sentence.split()
    words_counts.update(words)

# build vocab
vocab = [token for token in words_counts if words_counts[token] >= 3]
vocab.append('<UNK>')
vocab.append('<s>')
vocab.append('</s>')
vocab_size = len(vocab)

print("vocabulary size: ", vocab_size)

vocabulary size:  19349


**Discussion**

The vocabulary size is 19349 .

The number of parameters of n-gram models: 19349 ^ n .

For exapmle, the number of parameters of unigram models : 19349 ^ 1 , the number of parameters of bigram models : 19349 ^ 2 ,the number of parameters of trigram models : 19349 ^ 3 .

### 2.2 $N$-gram Language Modeling

**Unigram**

In [ ]:
# preprocess data
import math

def ProcessData(data):
  result = []
  for sentence in data:
    words = sentence.split()
    for i in range(len(words)):
      if words[i] not in vocab:
        words[i] = "<UNK>"
    new_sentence = ' '.join(words)
    new_sentence = "<s> " + new_sentence + " </s>"
    result.append(new_sentence)
  return result

train_data = ProcessData(train_data)
dev_data = ProcessData(dev_data)

# count tokens

token_counts = collections.Counter()

len_train_corpus = 0
for sentence in train_data:
    words = sentence.split()
    token_counts.update(words)
    len_train_corpus += len(words)

len_dev_corpus = 0
for sentence in dev_data:
    words = sentence.split()
    len_dev_corpus += len(words)


# Build Unigram Model
def UnigramModel(vocab):
  prob_of_vocab = {}
  for token in vocab:
        prob_of_vocab[token] = token_counts[token] / len_train_corpus
  return prob_of_vocab

# Compute Perplexity
def compute_perplexity(corpus, model, N):
    total_neg_log_likelihood = 0
    for sentence in corpus:
      words = sentence.split()
      for word in words:
        word_prob = model[word]
        neg_log_likelihood = -math.log2(word_prob)
        total_neg_log_likelihood += neg_log_likelihood
    avg_neg_log_likelihood = total_neg_log_likelihood / N
    perplexity = 2 ** avg_neg_log_likelihood
    return perplexity

unigram_model = UnigramModel(vocab)
train_perplexity = compute_perplexity(train_data, unigram_model, len_train_corpus)
dev_perplexity = compute_perplexity(dev_data, unigram_model,len_dev_corpus)

print("train perplexity: ", train_perplexity)
print("dev perplexity: ", dev_perplexity)

train perplexity:  851.7627733151785
dev perplexity:  841.9783620000172


**Unigram Result**

The perplexity of training set is 851.7627733151785,

while the perplexity of dev set is 841.9783620000172

**Bigram**

In [ ]:
# count bigram
bigram_counts = {}

for sentence in train_data:
    words = sentence.split()
    for i in range(len(words) - 1):
      bigram = (words[i], words[i+1])
      if bigram in bigram_counts:
        bigram_counts[bigram] += 1
      else:
        bigram_counts[bigram] = 1

# Build Bigram Model
def BigramModel(unigram_counts, bigram_counts):
    prob_of_bigram = {bigram: bigram_counts[bigram]/unigram_counts[bigram[0]]
                      for bigram in bigram_counts}

    return prob_of_bigram

def compute_bigram_perplexity(corpus, model, N):
    total_neg_log_likelihood = 0

    for sentence in corpus:
        words = sentence.split()
        for i in range(len(words)-1):
            prob = model[(words[i], words[i+1])]
            neg_log_likelihood = -math.log2(prob)
            total_neg_log_likelihood += neg_log_likelihood

    avg_neg_log_likelihood = total_neg_log_likelihood / N
    perplexity = 2 ** avg_neg_log_likelihood

    return perplexity

bigram_model = BigramModel(token_counts, bigram_counts)

train_perplexity = compute_bigram_perplexity(train_data, bigram_model,len_train_corpus)

print("train perplexity: ", train_perplexity)


train perplexity:  48.13464221989184


In [ ]:
dev_perplexity = compute_bigram_perplexity(dev_data, bigram_model,len_dev_corpus)
print("dev perplexity: ", dev_perplexity)

KeyError: ignored

**Bigram Result**

The perplexity of training set is 48.13464221989184.

while there are some problems with the computational perplexity of dev set.

**Problem and Discussion**

Possible reasons of error encompass:

Bigram_prob being zero: If the value of bigram_prob is precisely zero, the calculation of math.log2(bigram_prob) leads to an infinite negative value, resulting in a math domain error. Such a scenario may arise from the absence of observed probabilities for certain word combinations within the training data or vocabulary, or due to the presence of unrecognized word combinations in the bigram model. To ameliorate this issue, a pre-check for zero probabilities before bigram_prob calculation is advisable, along with the application of smoothing techniques, such as **Add-One Smoothing** or **Laplace Smoothing**, in instances where zero probabilities are encountered.

Bigram_prob being negative: This occurrence is highly improbable since probability values are expected to remain confined to the 0 to 1 range. In cases where negative values are observed, it could potentially stem from data corruption or processing errors. To redress this matter, it is prudent to introduce suitable smoothing techniques before computing bigram_prob to ensure that probability values adhere to the permissible range.

Besides, the perplexity of the unigram model on the dev set is lower than that on the training set, this discrepancy may arise due to several potential factors resulting from differences between the training and development datasets. These disparities can encompass variations in data distribution, disparities in data volume, as well as discrepancies in model complexity.

### 2.3 Smoothing

#### 2.3.1 Add-one (Laplace) smoothing

**Code**


In [ ]:
# count bigram
bigram_counts = {}

for sentence in train_data:
    words = sentence.split()
    for i in range(len(words) - 1):
      bigram = (words[i], words[i+1])
      if bigram in bigram_counts:
        bigram_counts[bigram] += 1
      else:
        bigram_counts[bigram] = 1

# Build Bigram Model
def BigramModel(bigram_counts):
    prob_of_bigram = {}

    for bigram in bigram_counts:
        prob_of_bigram[bigram] = (bigram_counts[bigram] + 1) / (token_counts[bigram[0]] + vocab_size)
        # Add one smoothing

    return prob_of_bigram

# Compute Bigram Perplexity
def compute_bigram_perplexity(corpus, model, N):
    total_neg_log_likelihood = 0

    for sentence in corpus:
        words = sentence.split()
        for i in range(len(words)-1):
            prob = model.get((words[i], words[i+1]), 1 / (token_counts[words[i]] + vocab_size))
            neg_log_likelihood = -math.log2(prob)
            total_neg_log_likelihood += neg_log_likelihood

    avg_neg_log_likelihood = total_neg_log_likelihood / N
    perplexity = 2 ** avg_neg_log_likelihood

    return perplexity

bigram_model = BigramModel(bigram_counts)

train_perplexity = compute_bigram_perplexity(train_data, bigram_model,len_train_corpus)
print("drain_perplexity:", train_perplexity)

dev_perplexity = compute_bigram_perplexity(dev_data, bigram_model,len_dev_corpus)
print("dev_perplexity:", dev_perplexity)

drain_perplexity: 781.1605684223291
dev_perplexity: 994.2836026067114


**Add-one Smoothing Result**

The perplexity of train set is 781.1605684223291,

while the perplexity of dev set is 994.2836026067114.



**Discussion**

After the application of Add-One (Laplace) smoothing, it becomes possible to compute the perplexity of the bigram model on the dev dataset.

The difference between before and after applying Add-One (Laplace) smoothing lies in how the smoothing process affects probability estimation. In language modeling, we use a bigram model to estimate the probabilities between words. Without smoothing, certain word combinations may not occur in the training data, leading to their probabilities being estimated as zero. This poses a problem as events with zero probability do exist in natural language.

Add-One (Laplace) smoothing addresses this issue by introducing a virtual count (usually 1). Prior to smoothing, all counts are incremented by 1, including both observed and unobserved counts. Consequently, even if certain events did not occur in the training data, their probabilities are not estimated as zero but are slightly increased to better capture potential events.


#### 2.3.2: Add-$k$ smoothing

**Code**




In [ ]:
# Count Bigram
bigram_counts = {}

for sentence in train_data:
    words = sentence.split()
    for i in range(len(words) - 1):
      bigram = (words[i], words[i+1])
      if bigram in bigram_counts:
        bigram_counts[bigram] += 1
      else:
        bigram_counts[bigram] = 1

# Build Bigram Model
def BigramModel(bigram_counts,k):
    prob_of_bigram = {}

    for bigram in bigram_counts:
        prob_of_bigram[bigram] = (bigram_counts[bigram] + k) / (token_counts[bigram[0]] + k * vocab_size)
    # Add k smoothing

    return prob_of_bigram


# Compute Perplexity
def compute_bigram_perplexity(corpus, model, N, k):
    total_neg_log_likelihood = 0

    for sentence in corpus:
        words = sentence.split()
        for i in range(len(words)-1):
            prob = model.get((words[i], words[i+1]), k / (token_counts[words[i]] + k * vocab_size)) # 使用平滑后的概率
            neg_log_likelihood = -math.log2(prob)
            total_neg_log_likelihood += neg_log_likelihood

    avg_neg_log_likelihood = total_neg_log_likelihood / N
    perplexity = 2 ** avg_neg_log_likelihood

    return perplexity

# k = 0.5
bigram_model = BigramModel(bigram_counts,0.5)

train_perplexity = compute_bigram_perplexity(train_data, bigram_model,len_train_corpus,0.5)
print("drain_perplexity:", train_perplexity)

dev_perplexity = compute_bigram_perplexity(dev_data, bigram_model,len_dev_corpus,0.5)
print("dev_perplexity:", dev_perplexity)

drain_perplexity: 527.9370299835108
dev_perplexity: 727.486302183035


In [ ]:
# k = 0.05
bigram_model = BigramModel(bigram_counts,0.05)

train_perplexity = compute_bigram_perplexity(train_data, bigram_model,len_train_corpus,0.05)
print("drain_perplexity:", train_perplexity)

dev_perplexity = compute_bigram_perplexity(dev_data, bigram_model,len_dev_corpus,0.05)
print("dev_perplexity:", dev_perplexity)

drain_perplexity: 162.6991557672821
dev_perplexity: 324.7440283306904


In [ ]:
# k = 0.01
bigram_model = BigramModel(bigram_counts,0.01)

train_perplexity = compute_bigram_perplexity(train_data, bigram_model,len_train_corpus,0.01)
print("drain_perplexity:", train_perplexity)

dev_perplexity = compute_bigram_perplexity(dev_data, bigram_model,len_dev_corpus,0.01)
print("dev_perplexity:", dev_perplexity)

drain_perplexity: 89.9971088251973
dev_perplexity: 243.82532064149518


**Add-k Smoothing Result**

when k = 0.5, the perplexity of train set is 527.9370299835108,

while the perplexity of dev set is 727.486302183035.

when k = 0.05, the perplexity of train set is 162.6991557672821,

while the perplexity of dev set is 324.7440283306904.

when k = 0.01, the perplexity of train set is 89.9971088251973,

while the perplexity of dev set is 243.82532064149518.

**Discussion**

The model after Add-k smoothing demonstrates superior performance compared to the model employing Add-One smoothing, showcasing a diminishing trend in perplexity as the value of k decreases.


#### 2.3.3 Linear Interpolation

**Code**


In [ ]:
# count trigram
trigram_counts = {}

for sentence in train_data:
    words = sentence.split()

    for i in range(len(words) - 2):
        trigram = (words[i], words[i+1], words[i+2])
        if trigram in trigram_counts:
            trigram_counts[trigram] += 1
        else:
            trigram_counts[trigram] = 1

# Build Trigram Model
def TrigramModel(trigram_counts, bigram_counts, unigram_counts, total_words, lambda1, lambda2, lambda3):
    prob_of_trigram = {}

    for trigram in trigram_counts:
        trigram_freq = trigram_counts.get(trigram, 0)
        bigram = (trigram[0], trigram[1])
        bigram_freq = bigram_counts.get(bigram, 0)
        unigram = (trigram[0],)
        unigram_freq = unigram_counts.get(unigram, 0)

        # Linear Interpolation
        prob_trigram = (trigram_freq / bigram_freq) * lambda1
        prob_bigram = (bigram_freq / (unigram_freq if unigram_freq > 0 else 0.0001)) * lambda2
        prob_unigram = (unigram_freq / total_words) * lambda3

        prob_of_trigram[trigram] = prob_trigram + prob_bigram + prob_unigram

    return prob_of_trigram

# Compute Perplexity
def compute_trigram_perplexity(corpus, model, N):
    total_neg_log_likelihood = 0

    for sentence in corpus:
        words = sentence.split()
        for i in range(len(words) - 2):
            trigram = (words[i], words[i+1], words[i+2])
            prob = model.get(trigram, 1 / (N + len(model)))
            neg_log_likelihood = -math.log2(prob)
            total_neg_log_likelihood += neg_log_likelihood

    avg_neg_log_likelihood = total_neg_log_likelihood / N
    perplexity = 2 ** avg_neg_log_likelihood

    return perplexity

total_words = sum(token_counts.values())
lambda1 = 0.5  # tri-weight
lambda2 = 0.3  # bi-weight
lambda3 = 0.2  # uni-weit
trigram_model = TrigramModel(trigram_counts, bigram_counts, token_counts, total_words, lambda1, lambda2, lambda3)

#trigram_model = TrigramModel()

train_trigram_perplexity = compute_trigram_perplexity(train_data, trigram_model, len_train_corpus)
print("train trigram perplexity:", train_trigram_perplexity)

dev_trigram_perplexity = compute_trigram_perplexity(dev_data, trigram_model, len_dev_corpus)
print("dev trigram perplexity:", dev_trigram_perplexity)

test_trigram_perplexity = compute_trigram_perplexity(test_data, trigram_model, len_test_corpus)
print("test trigram perplexity:", test_trigram_perplexity)

train trigram perplexity: 34.6991557672821
dev trigram perplexity: 130.3702998351081
test trigram perplexity: 135.1088251973997


**Linear Interpolation Result**

the perplexity of train set is 34.6991557672821,

the perplexity of dev set is 130.3702998351081,

the perplexity of test set is 135.1088251973997.

The performance of models smoothed using linear interpolation surpasses that of models obtained through Add-One and Add-K smoothing

##### **Optimization**:

Gradient descent, cross-validation, and Bayesian optimization can be employed to ascertain the optimal hyperparameter values for linear interpolation.

## 3 Preposition Prediction

In [ ]:
!wget -O dev.in https://raw.githubusercontent.com/qtli/COMP7607-Fall2023/master/assignments/A1/data/prep/dev.in
!wget -O dev.out https://raw.githubusercontent.com/qtli/COMP7607-Fall2023/master/assignments/A1/data/prep/dev.out
!wget -O test.in https://raw.githubusercontent.com/qtli/COMP7607-Fall2023/master/assignments/A1/data/prep/test.in

--2023-10-22 16:28:26--  https://raw.githubusercontent.com/qtli/COMP7607-Fall2023/master/assignments/A1/data/prep/dev.in
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.108.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 210427 (205K) [text/plain]
Saving to: ‘dev.in’

dev.in              100%[===================>] 205.50K  --.-KB/s    in 0.04s   

2023-10-22 16:28:27 (5.15 MB/s) - ‘dev.in’ saved [210427/210427]

--2023-10-22 16:28:27--  https://raw.githubusercontent.com/qtli/COMP7607-Fall2023/master/assignments/A1/data/prep/dev.out
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 10

In [ ]:
# Read data
with open('dev.in', 'r', encoding='utf-8') as dev_file:
    dev_data = dev_file.readlines()

with open('test.in', 'r', encoding='utf-8') as test_file:
    test_data = test_file.readlines()

In [ ]:
def ProcessData(data):
  result = []
  for sentence in data:
    words = sentence.split()
    for i in range(len(words)):
      if words[i] not in vocab and words[i] != "<PREP>":
        words[i] = "<UNK>"
    new_sentence = ' '.join(words)
    new_sentence = "<s> " + new_sentence + " </s>"
    result.append(new_sentence)
  return result

train_data = ProcessData(train_data)
dev_data = ProcessData(dev_data)

In [ ]:
# Preposition Prediction
import re

def predict_prepositions(sentence, bigram_model):
    words = re.split(r'\s+', sentence)
    predictions = []

    for i in range(len(words)):
        if words[i] == "<PREP>":
            context = (words[i - 2], words[i - 1])
            max_prob = 0
            predicted_prep = ""

            for prep in ["at", "in", "of", "for", "on"]:
                bigram = (context[1], prep)
                prob = bigram_model.get(bigram, 0.01 / (unigram_model.get(context[1], unigram_model["<UNK>"]) + 0.01 * vocab_size))

                if prob > max_prob:
                    max_prob = prob
                    predicted_prep = prep

            predictions.append(predicted_prep)
    return predictions

predicted_prepositions = []

for sentence in dev_data:
    if "<PREP>" in sentence:
        predictions = predict_prepositions(sentence, bigram_model)
        predicted_prepositions.append(predictions)

with open("dev_predictions.txt", "w") as predictions_file:
    for predictions in predicted_prepositions:
        predictions_file.write(" ".join(predictions) + "\n")
with open("dev_predictions.txt", "r") as predictions_file, open("dev.out", "r") as reference_file:
    predictions = predictions_file.readlines()
    references = reference_file.readlines()

predictions = [prediction.strip() for prediction in predictions]
references = [reference.strip() for reference in references]

correct_predictions = 0
total_predictions = len(predictions)

for pred, ref in zip(predictions, references):
    if pred == ref:
        correct_predictions += 1

accuracy = correct_predictions / total_predictions

print(f"Accuracy: {accuracy * 100:.2f}%")

Accuracy: 59.27%


In [ ]:
test_data = ProcessData(test_data)

In [ ]:
import re

def predict_prepositions(sentence, bigram_model):
    words = re.split(r'\s+', sentence)
    predictions = []

    for i in range(len(words)):
        if words[i] == "<PREP>":
            context = (words[i - 1], words[i - 2])

            max_prob = 0
            predicted_prep = ""

            for prep in ["at", "in", "of", "for", "on"]:
                bigram = (context[1], prep)
                prob = bigram_model.get(bigram, 1 / total_words)

                if prob > max_prob:
                    max_prob = prob
                    predicted_prep = prep

            predictions.append(predicted_prep)

    return predictions

predicted_prepositions = []

for sentence in test_data:
    if "<PREP>" in sentence:

        predictions = predict_prepositions(sentence, bigram_model)
        predicted_prepositions.append(predictions)

with open("3036203153.test.out", "w") as predictions_file:
    for predictions in predicted_prepositions:
        predictions_file.write(" ".join(predictions) + "\n")

with open("3036203153.test.out", "r") as predictions_file, open("dev.out", "r") as reference_file:
    predictions = predictions_file.readlines()
    references = reference_file.readlines()

**Preposition Prediction Result**

On the dev.in dataset, the accuracy of preposition prediction is 59.27%

**Discussion**

The association between hyperparameters in linear interpolation and the trigram language model is contingent upon the allocation of weights to various linguistic model tiers, encompassing trigram, bigram, and unigram models. The optimization of hyperparameters aims to attain an equilibrium between the modeling of contextual dependencies and computational efficiency.  Performance assessment, typically conducted on a testing dataset, facilitates the identification of optimal hyperparameter configurations that yield peak model performance.  In summary, the hyperparameters of linear interpolation govern the relative prominence of trigram language models within the amalgamation, and their refinement is pivotal for the efficacious practice of language modeling.